### DISCUSSION TITLE
Peak to gene links - restrict to TSS links?

### DISCUSSION THREAD

**Discussion by jblich870** [GreenleafLab/ArchR#1127] [Questions / Documentation] [answered]
Hi,

I've run addPeak2GeneLinks followed by getPeak2GeneLinks with returnLoops = T, and have attached a representative track below. 

[Plot.pdf](https://github.com/GreenleafLab/ArchR/files/7453850/Plot.pdf)

Many of the displayed 'peak2genelinks' do not include the TSS - instead of linking peaks to genes, they appear to be just linking peaks.

Why is this? And is there a way to restrict the links so they only link peaks to genes?

Many thanks,
J

**Comment by rcorces** [comment, marked as answer]
This shouldnt be the case. Can you provide a reproducible example using the tutorial data? My assumption is that there is something off about your geneAnnotation

**Reply by jblich870** [reply]
I am having difficulty with the tutorial data - I have similar issues with creating Arrow files as described in #1123 and #1104.

However, I don't think there is a problem with gene annotation, as at other loci the links appear to be correct:

[Plot1.pdf](https://github.com/GreenleafLab/ArchR/files/7459331/Plot1.pdf)
[Plot2.pdf](https://github.com/GreenleafLab/ArchR/files/7459330/Plot2.pdf)

It is only at certain loci where there seem to be weird links, eg:

[Plot3.pdf](https://github.com/GreenleafLab/ArchR/files/7459332/Plot3.pdf)
[Plot4.pdf](https://github.com/GreenleafLab/ArchR/files/7459333/Plot4.pdf)

My dataset is 10X Multiome, and I'm using hg38. Code used for links:

Multiproj <- addPeak2GeneLinks(
  ArchRProj = Multiproj,
  reducedDims = "LSI_Combined",
  useMatrix = "GeneExpressionMatrix",
  cellsToUse = NULL,
  addEmpiricalPval = T,
  maxDist = 500000)

p2g.LOOPS <- getPeak2GeneLinks(
  ArchRProj = Multiproj,
  corCutOff = 0.45,
  varCutOffATAC = 0.25,
  varCutOffRNA = 0.25,
  resolution = 1,
  returnLoops = T)

p1 <- plotBrowserTrack(
  ArchRProj = Multiproj, 
  groupBy = "Clusters", 
  geneSymbol = "SNAI2", 
  upstream = 10000,
  downstream = 10000,
  loops = p2g.LOOPS)

plotPDF(plotList = p1, name = "Plot1.pdf", ArchRProj = Multiproj, addDOC = FALSE, width = 5, height = 5)

Let me know if you need anything else. 
Thanks,
J

**Reply by rcorces** [reply]
I think I need a reproducible example with the tutorial dataset to actually understand what is going on. There shouldnt be problems creating arrows from the tutorial data if the downloads are successful, assuming you are using `release_1.0.1` or the most up-to-date `release_1.0.2`.

**Reply by jblich870** [reply]
I actually think the problem is not with creating arrow files but with opening something required for addGeneIntegrationMatrix (below). I get a similar error for addGroupCoverages.


devtools::install_github("GreenleafLab/ArchR", 
                         ref="release_1.0.2", repos = BiocManager::repositories())
library(ArchR)
addArchRThreads(threads = 8) 
addArchRGenome("hg19")
setwd("~/Desktop/ArchR_tutorial/")

######

inputFiles <- getTutorialData("Hematopoiesis")

ArrowFiles <- createArrowFiles(
  inputFiles = inputFiles,
  sampleNames = names(inputFiles),
  filterTSS = 4,
  filterFrags = 1000, 
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

doubScores <- addDoubletScores(
  input = ArrowFiles,
  k = 10, 
  knnMethod = "UMAP", 
  LSIMethod = 1
)

projHeme1 <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "HemeTutorial",
  copyArrows = TRUE
)


saveArchRProject(ArchRProj = projHeme1, outputDirectory = "Save-ProjHeme1", load = FALSE)


projHeme2 <- filterDoublets(projHeme1)

projHeme2 <- addIterativeLSI(
  ArchRProj = projHeme2,
  useMatrix = "TileMatrix", 
  name = "IterativeLSI", 
  iterations = 2, 
  clusterParams = list(
    resolution = c(0.2), 
    sampleCells = 10000, 
    n.start = 10
  ), 
  varFeatures = 25000, 
  dimsToUse = 1:30
)

projHeme2 <- addClusters(
  input = projHeme2,
  reducedDims = "IterativeLSI",
  method = "Seurat",
  name = "Clusters",
  resolution = 0.8
)

projHeme2 <- addUMAP(
  ArchRProj = projHeme2, 
  reducedDims = "IterativeLSI", 
  name = "UMAP", 
  nNeighbors = 30, 
  minDist = 0.5, 
  metric = "cosine"
)

p1 <- plotEmbedding(ArchRProj = projHeme2, colorBy = "cellColData", name = "Sample", embedding = "UMAP")
p2 <- plotEmbedding(ArchRProj = projHeme2, colorBy = "cellColData", name = "Clusters", embedding = "UMAP")
ggAlignPlots(p1, p2, type = "h")

projHeme2 <- addImputeWeights(projHeme2)

if(!file.exists("scRNA-Hematopoiesis-Granja-2019.rds")){
  download.file(
    url = "https://jeffgranja.s3.amazonaws.com/ArchR/TestData/scRNA-Hematopoiesis-Granja-2019.rds",
    destfile = "scRNA-Hematopoiesis-Granja-2019.rds"
  )
}

seRNA <- readRDS("scRNA-Hematopoiesis-Granja-2019.rds")

projHeme2 <- addGeneIntegrationMatrix(
  ArchRProj = projHeme2, 
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI",
  seRNA = seRNA,
  addToArrow = FALSE,
  groupRNA = "BioClassification",
  nameCell = "predictedCell_Un",
  nameGroup = "predictedGroup_Un",
  nameScore = "predictedScore_Un"
)

cM <- as.matrix(confusionMatrix(projHeme2$Clusters, projHeme2$predictedGroup_Un))
preClust <- colnames(cM)[apply(cM, 1 , which.max)]
cbind(preClust, rownames(cM)) #Assignments
unique(unique(projHeme2$predictedGroup_Un))

cTNK <- paste0(paste0(19:25), collapse="|")

cNonTNK <- paste0(c(paste0("0", 1:9), 10:13, 15:18), collapse="|")

clustTNK <- rownames(cM)[grep(cTNK, preClust)]

clustNonTNK <- rownames(cM)[grep(cNonTNK, preClust)]

rnaTNK <- colnames(seRNA)[grep(cTNK, colData(seRNA)$BioClassification)]

rnaNonTNK <- colnames(seRNA)[grep(cNonTNK, colData(seRNA)$BioClassification)]

groupList <- SimpleList(
  TNK = SimpleList(
    ATAC = projHeme2$cellNames[projHeme2$Clusters %in% clustTNK],
    RNA = rnaTNK
  ),
  NonTNK = SimpleList(
    ATAC = projHeme2$cellNames[projHeme2$Clusters %in% clustNonTNK],
    RNA = rnaNonTNK
  )    
)

saveArchRProject(ArchRProj = projHeme2, outputDirectory = "Save-ProjHeme2", load = FALSE)


projHeme3 <- addGeneIntegrationMatrix(
  ArchRProj = projHeme2, 
  useMatrix = "GeneScoreMatrix",
  matrixName = "GeneIntegrationMatrix",
  reducedDims = "IterativeLSI",
  seRNA = seRNA,
  addToArrow = T,
  force= TRUE,
  groupList = groupList,
  groupRNA = "BioClassification",
  nameCell = "predictedCell",
  nameGroup = "predictedGroup",
  nameScore = "predictedScore")




ArchR logging to : ArchRLogs/ArchR-addGeneIntegrationMatrix-17ce1215311bf-Date-2021-11-03_Time-10-46-00.log
If there is an issue, please report to github with logFile!
2021-11-03 10:46:00 : Running Seurat's Integration Stuart* et al 2019, 0.002 mins elapsed.
2021-11-03 10:46:00 : Checking ATAC Input, 0.013 mins elapsed.
2021-11-03 10:46:00 : Checking RNA Input, 0.013 mins elapsed.
2021-11-03 10:46:07 : Found 18601 overlapping gene names from gene scores and rna matrix!, 0.126 mins elapsed.
2021-11-03 10:46:07 : Creating Integration Blocks, 0.126 mins elapsed.
2021-11-03 10:46:07 : Prepping Interation Data, 0.129 mins elapsed.
2021-11-03 10:46:08 : Computing Integration in 2 Integration Blocks!, 0 mins elapsed.
2021-11-03 10:50:54 : Block (1 of 2) : Plotting Joint UMAP, 4.762 mins elapsed.
2021-11-03 10:51:09 : Block (2 of 2) : Plotting Joint UMAP, 5.017 mins elapsed.
2021-11-03 10:51:28 : Transferring Data to ArrowFiles, 5.331 mins elapsed.
Error in .safelapply(seq_along(allSamples), function(y) { : 
Error Found Iteration 2 : 
	[1] "Error in H5Fopen(file) : HDF5. File accessibility. Unable to open file.\n"
	<simpleError in H5Fopen(file): HDF5. File accessibility. Unable to open file.>
Error Found Iteration 3 : 
	[1] "Error in H5Fopen(file) : HDF5. File accessibility. Unable to open file.\n"
	<simpleError in H5Fopen(file): HDF5. File accessibility. Unable to open file.>
In addition: Warning message:
In mclapply(..., mc.cores = threads, mc.preschedule = preschedule) :
  2 function calls resulted in an error

**Reply by rcorces** [reply]
maybe your problem is related to this - https://github.com/GreenleafLab/ArchR/issues/248

**Reply by jblich870** [reply]
Thanks for the suggestion and apologies for the delayed reply. I have got the tutorial working, and the links seem to be fine - as far as I can tell they all link to a TSS.

I can't figure out why I have weird links in my dataset. The vast majority of links appear to be fine and link to a TSS, but many don't. Any ideas?

Would be very grateful if you could help me figure this out. Let me know if you need anything.

Many thanks,
J

**Reply by rcorces** [reply]
it would be very challenging for me to provide substantive help without a reproducible example. From the code, it would be surprising to me if you have links being predicted that map to regions that are not TSSs as defined in your geneAnnotation.

**Reply by jblich870** [reply]
So do you think the geneAnnotation is wrong somehow? How could I check / fix this? 

I've had a look at the output from 

getArchRGenome(geneAnnotation = T, genomeAnnotation = FALSE)

but not really sure what to look for.

Any advice would be appreciated.
Thanks,
J

**Reply by rcorces** [reply]
If you are using one of the ArchR-provided gene annotations (i.e. you didnt create your own) then I would not expect this to be a problem. You can see all of the TSSs in your geneAnnotation using:
```
> gA <- getGeneAnnotation(projHeme5)
> gA
List of length 3
names(3): genes exons TSS
> gA$TSS
GRanges object with 49052 ranges and 2 metadata columns:
          seqnames    ranges strand |     tx_id     tx_name
             <Rle> <IRanges>  <Rle> | <integer> <character>
      [1]     chr1     11874      + |         1  uc001aaa.3
      [2]     chr1     69091      + |         4  uc001aal.1
      [3]     chr1    321084      + |         5  uc001aaq.2
      [4]     chr1    321146      + |         6  uc001aar.2
      [5]     chr1    322037      + |         7  uc009vjk.2
      ...      ...       ...    ... .       ...         ...
  [49048]     chrY  27605678      - |     78803  uc004fwx.1
  [49049]     chrY  27606421      - |     78804  uc022cpc.1
  [49050]     chrY  27607432      - |     78805  uc004fwz.3
  [49051]     chrY  27635954      - |     78806  uc022cpd.1
  [49052]     chrY  59360854      - |     78807  uc011ncc.1
  -------
  seqinfo: 24 sequences from hg19 genome
> 
```

**Reply by jblich870** [reply]
I'm using ArchR-provided hg38:

addArchRGenome("hg38")

Using your code for my project gives the below. Seems odd that there are 166023 TSSs, versus only 49052 with hg19?


> gA <- getGeneAnnotation(Multiproj)
> gA
List of length 3
names(3): genes exons TSS

> gA$TSS
GRanges object with 166023 ranges and 2 metadata columns:
           seqnames    ranges strand |     tx_id     tx_name
              <Rle> <IRanges>  <Rle> | <integer> <character>
       [1]     chr1     29554      + |         1  uc057aty.1
       [2]     chr1     30267      + |         2  uc057atz.1
       [3]     chr1     30366      + |         3  uc031tlb.1
       [4]     chr1     69091      + |         4  uc001aal.1
       [5]     chr1    160446      + |         5  uc057aum.1
       ...      ...       ...    ... .       ...         ...
  [166019]     chrY  25394719      - |    182394  uc065cwf.1
  [166020]     chrY  25391610      - |    182395  uc065cwg.1
  [166021]     chrY  25498059      - |    182396  uc065cwl.1
  [166022]     chrY  26411158      - |    182397  uc065cwp.1
  [166023]     chrY  57203357      - |    182398  uc065cww.1
  -------
  seqinfo: 24 sequences from hg38 genome

**Reply by rcorces** [reply]
This is expected as genome annotations continue to improve over time. hg38 has better transcript annotations.

I think the point that hasnt been conveyed is that one gene can have many transcriptional start sites. In the hg38 `TxDb` object, these alternative start sites are annotated and thus they are a part of the ArchRProject. In ArchR, only one ___gene___ is plotted in the browser track for cleanliness but there may be many ___transcripts___ for that gene, some of which might have alternative TSSs. So just because you dont see the different gene tracks for different transcripts, doesnt mean that they dont exist. Perhaps this is what you were observing previously when you said that your links didnt map to TSSs.

**Reply by jblich870** [reply]
Right, I hadn’t considered that - that helps to clear things up a bit, thanks.

It does however raise a couple of separate, but related, issues:

First is that some genes seem to be absent from the browser track whereas their TSSs are annotated in the hg38 object. This leads to the peculiar situation of having links in the track for invisible genes (e.g. AC016745.3 in the attached plots).

[Plot 5.pdf](https://github.com/GreenleafLab/ArchR/files/7579914/Plot.5.pdf)
<img width="892" alt="Plot 6" src="https://user-images.githubusercontent.com/69856293/142841185-67db191e-e539-4975-be0a-878dacd205b1.png">

Second is that some links appear to be between the promoter of one gene and the TSS of a neighbouring gene (eg below). This is probably because genes in close proximity on the genome can covary in expression, and so size of the promoter will correlate with expression of the neighbouring gene even though there is unlikely to be a biological link between the two. Is there a way to filter out such links?

[Plot 7.pdf](https://github.com/GreenleafLab/ArchR/files/7579917/Plot.7.pdf)

Additionally, it might be helpful to be able to selectively visualise links for only one gene / one TSS in the browser track. Many links are long range, but in order to capture these in the track you end up getting lots of links for nearby genes and the plot can look messy.

**Reply by rcorces** [reply]
My general response is: Gene and transcript annotation is a complicated problem that we (ArchR) are not wading into and we have entirely off-loaded this problem to the end user and TxDb. ArchR will use whatever gene annotation you give it. If you dont like your gene annotation, change it to your liking. But ArchR isnt a gene annotation package so we steer clear of this entirely.

> First is that some genes seem to be absent from the browser track whereas their TSSs are annotated in the hg38 object. This leads to the peculiar situation of having links in the track for invisible genes (e.g. AC016745.3 in the attached plots).

As mentioned above, this falls into the "if you dont like the default, make your own". 

> Second is that some links appear to be between the promoter of one gene and the TSS of a neighbouring gene (eg below). This is probably because genes in close proximity on the genome can covary in expression, and so size of the promoter will correlate with expression of the neighbouring gene even though there is unlikely to be a biological link between the two. 

Its all about correlation. There is no true implication of causation and there are examples in biology of the promoter of one gene "regulating" the expression of another gene.

> Is there a way to filter out such links?

Sure, but this is up to the end-user. Filter the returned GRanges object however you want.

> Additionally, it might be helpful to be able to selectively visualise links for only one gene / one TSS in the browser track. Many links are long range, but in order to capture these in the track you end up getting lots of links for nearby genes and the plot can look messy.

You can add arbitrary tracks to your browser visualizations. You can make whatever GRanges object you want by filtering as mentioned above and add it as a `loopTrack` to `plotBrowserTrack()` or specify the `loops` argument to `ArchRBrowser()`.

**Reply by jblich870** [reply]
Fair enough, and good points/suggestions. 
Really appreciate your help, thanks.
J
